### ISSUE TITLE
Clean up garbage characters in MS MARCO dataset

### ISSUE THREAD

**Issue by lintool** [castorini/anserini#945]
I've noticed that MS MARCO data has garbage characters. A bit of probing, appears to be MS-related encoding issues: https://stackoverflow.com/questions/2477452/%C3%A2%E2%82%AC-showing-on-page-instead-of

Let's try some cleanup and see if it improves effectiveness?

**Comment by edwinzhng** [comment]
Does this apply to the MS MARCO Document Ranking dataset as well? I looked at some of the indexed documents and the characters seem to be encoded correctly

**Comment by lintool** [comment]
Okay, here's a concrete example:

```
{"id": "7287406", "contents": "Why do Bears hibernate? March 31, 2010, Joan, Leave a comment. Why do bears hibernate? When we hear the word \u00e2\u0080\u0098hibernate\u00e2\u0080\u0099 we always associate it with bears. That is because, while a lot of animals go through hibernation during the winter season such as squirrels, rodents and bats, the bear is the most famous when it comes to hibernating. What comes first to our mind is why do bears hibernate? First of all, let\u00e2\u0080\u0099s get to know the meaning of hibernation."}
```

In browser, it gets rendered as:

<img width="876" alt="Screen Shot 2020-01-13 at 7 54 11 PM" src="https://user-images.githubusercontent.com/313837/72304424-7fe7ee80-363e-11ea-9139-9bb561d10600.png">

Note the `â` garbage.

**Comment by edwinzhng** [comment]
Sorry for all the back and forth here, will try to summarize what I found while digging around (let's work with the single example above but it should extend to others as well):

### TLDR
Using Latin-1 should work since the underlying byte values are the same as Windows-1252. Also Python is unable to encode characters between `\u0080` and `\u009f` using Windows-1252 since they are undefined - therefore we need to use Latin-1.

### Latin-1 vs Windows-1252

I found a [Stack Overflow answer](https://stackoverflow.com/questions/15271453/what-is-this-character-%c3%a2-u0080-u0099) indicating that `\u00e2\u0080\u0099` is from ISO-8859-1 which is the same as Latin-1
 
This is confusing because it does not match the original Stack Overflow you sent in the issue description that mentions Windows-1252. However, apparently the **Windows-1252 encoding is the exact same as Latin-1 encoding** except for the fact that the `0x80-0x9F` range maps to different characters.

Okay, now let's look at the raw bytes for the text:

```
b'7287406\tWhy do Bears hibernate? March 31, 2010, Joan, Leave a comment. Why do bears hibernate? When we hear the word \xc3\xa2\xc2\x80\xc2\x98hibernate\xc3\xa2\xc2\x80\xc2\x99 we always associate it with bears. That is because, while a lot of animals go through hibernation during the winter season such as squirrels, rodents and bats, the bear is the most famous when it comes to hibernating. What comes first to our mind is why do bears hibernate? First of all, let\xc3\xa2\xc2\x80\xc2\x99s get to know the meaning of hibernation.\n'
```

Notice that `\xc3\xa2\xc2\x80\xc2\x99`  converted to UTF-8 is  `\u00e2\u0080\u0099`. What likely happened is the original data was encoded properly as UTF-8, mistakenly decoded as Latin-1 or Windows-1252 when reading the text (creating garbage), and then encoded back as UTF-8 when saving (preserving the same garbage).

Reverting this is just the same two steps but reversed:

1. Decode the UTF-8, which should give Windows-1252 or Latin-1 encoded bytes
2. Encode the Windows-1252 or Latin-1 bytes back into UTF-8

Theoretically, since Windows-1252 and Latin-1 produce the exact same byte values from 0-255, both character sets should also decode the UTF-8 to the same values. This explains why `latin-1` decoded things properly in my PR. It's also is where the difference in the `0x80-0x9F` range comes into play. Apparently in Python, this range for `cp1252` is undefined and throws the error shown below, so it **cannot** encode the original bytes, whereas `latin-1` has the values defined properly - hence we need to use `latin-1`.

```
>>> '\u0080'.encode('cp1252')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python37\lib\encodings\cp1252.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_table)
```

vs 

```
>>> '\u0080'.encode('latin1')
b'\x80'
```

### Why Decoding Sometimes Fails

To explain why some decodings still fail even after the proper encoding/decoding (~46 lines in the passage dataset), we can look at this page (https://www.i18nqa.com/debug/bug-double-conversion.html) concerning a double mis-conversion from Windows-1252 to UTF-8. This states that for 5 specific characters (Á Í Ï Ð Ý), once the text has already been incorrectly converted from UTF-8 to another format, it is impossible to convert it back to it's valid form. Instead, the UTF-8 character will always be invalid.

Take the second row as an example here:

![image](https://user-images.githubusercontent.com/21136804/72312666-cfd2af80-3656-11ea-9b0b-cfea6b5fa7ac.png)

**Comment by arjenpdevries** [comment]
Thank you for the pointer to the double mis-conversion from Windows-1252 to UTF-8. I have seen that before but never understood!

**Comment by nsndimt** [comment]
I have seen people talked about this issue and here is a potential [fix using fyft](https://github.com/spacemanidol/MSMARCO/issues/26#issuecomment-488841050). 
```
In [8]: from ftfy import fix_text                                                                             

In [10]: print(content)                                                                                       
Why do Bears hibernate? March 31, 2010, Joan, Leave a comment. Why do bears hibernate? When we hear the word âhibernateâ we always associate it with bears. That is because, while a lot of animals go through hibernation during the winter season such as squirrels, rodents and bats, the bear is the most famous when it comes to hibernating. What comes first to our mind is why do bears hibernate? First of all, letâs get to know the meaning of hibernation.

In [11]: print(fix_text(content))                                                                             
Why do Bears hibernate? March 31, 2010, Joan, Leave a comment. Why do bears hibernate? When we hear the word 'hibernate' we always associate it with bears. That is because, while a lot of animals go through hibernation during the winter season such as squirrels, rodents and bats, the bear is the most famous when it comes to hibernating. What comes first to our mind is why do bears hibernate? First of all, let's get to know the meaning of hibernation.
```

**Comment by lintool** [comment]
Interesting, thanks @nsndimt for the link. The link above suggests that the issue has been fixed...? But I don't think it has? What gives?

@edwinzhng Can you please take a look?

**Comment by edwinzhng** [comment]
> Interesting, thanks @nsndimt for the link. The link above suggests that the issue has been fixed...? But I don't think it has? What gives?
> 
> @edwinzhng Can you please take a look?

Hmm, I can't really see why this shouldn't be fixed if they used `ftfy`. They mentioned in the issue that it was only done for top1000 dev and eval and the triples, which means `collection.tsv` could still have the same encoding errors. There was also another issue https://github.com/spacemanidol/MSMARCO/issues/25 that mentioned corrupted characters in `collection.tsv` but was closed without further action.

On that note, it would be good to include `ftfy` in the future to fix encoding issues instead of doing it manually like I did in https://github.com/castorini/anserini/pull/947.
